File: samples/src/main/java/tk/zielony/carbonsamples/component/AvatarTextListItemActivity.java
package tk.zielony.carbonsamples.component;

import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;

import carbon.component.AvatarTextRow;
import carbon.component.DefaultAvatarTextItem;
import carbon.component.DividerItem;
import carbon.component.DividerRow;
import carbon.component.PaddingItem;
import carbon.component.PaddingRow;
import carbon.recycler.RowListAdapter;
import carbon.widget.RecyclerView;
import tk.zielony.carbonsamples.ActivityAnnotation;
import tk.zielony.carbonsamples.R;
import tk.zielony.carbonsamples.ThemedActivity;
import tk.zielony.randomdata.Generator;
import tk.zielony.randomdata.RandomData;
import tk.zielony.randomdata.person.DrawableAvatarGenerator;
import tk.zielony.randomdata.person.StringNameGenerator;

@ActivityAnnotation(layout = R.layout.activity_listcomponent, title = R.string.avatarTextListItemActivity_title)
public class AvatarTextListItemActivity extends ThemedActivity {

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initToolbar();

        List<Serializable> items = Arrays.asList(
                new PaddingItem(getResources().getDimensionPixelSize(R.dimen.carbon_paddingHalf)),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DividerItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DividerItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DividerItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new DefaultAvatarTextItem(),
                new PaddingItem(getResources().getDimensionPixelSize(R.dimen.carbon_paddingHalf)));

        RandomData randomData = new RandomData();
        randomData.addGenerators(new Generator[]{
                new DrawableAvatarGenerator(this),
                new StringNameGenerator().withMatcher(f -> f.getName().equals("text"))
        });
        randomData.fill(items);

        RecyclerView recycler = findViewById(R.id.recycler);
        recycler.setLayoutManager(new LinearLayoutManager(this));

        RowListAdapter<Serializable> adapter = new RowListAdapter<>(DefaultAvatarTextItem.class, AvatarTextRow::new);
        adapter.addFactory(PaddingItem.class, PaddingRow::new);
        adapter.addFactory(DividerItem.class, DividerRow::new);
        recycler.setAdapter(adapter);
        adapter.setItems(items);
    }
}

File: samples/src/main/java/tk/zielony/carbonsamples/component/AvatarTextListItemActivity.kt
package tk.zielony.carbonsamples.component

import android.graphics.drawable.Drawable
import android.os.Bundle
import androidx.recyclerview.widget.LinearLayoutManager
import carbon.component.AvatarTextRow
import carbon.component.DefaultAvatarTextItem
import carbon.recycler.RowListAdapter
import carbon.recycler.ViewItemDecoration
import kotlinx.android.synthetic.main.activity_listcomponent.*
import tk.zielony.carbonsamples.R
import tk.zielony.carbonsamples.SampleAnnotation
import tk.zielony.carbonsamples.ThemedActivity
import tk.zielony.randomdata.RandomData
import tk.zielony.randomdata.Target
import tk.zielony.randomdata.person.DrawableAvatarGenerator
import tk.zielony.randomdata.person.StringNameGenerator

@SampleAnnotation(layoutId = R.layout.activity_listcomponent, titleId = R.string.avatarTextListItemActivity_title)
class AvatarTextListItemActivity : ThemedActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initToolbar()

        val items = listOf(
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem(),
                DefaultAvatarTextItem())

        val randomData = RandomData().apply {
            addGenerator(Drawable::class.java, DrawableAvatarGenerator(this@AvatarTextListItemActivity))
            addGenerator(String::class.java, StringNameGenerator().withMatcher { f: Target -> f.name == "text" })
        }
        randomData.fill(items)

        val adapter = RowListAdapter<DefaultAvatarTextItem>().apply {
            putFactory(DefaultAvatarTextItem::class.java, { AvatarTextRow(it) })
        }
        adapter.items = items

        recycler.layoutManager = LinearLayoutManager(this)
        recycler.adapter = adapter

        val paddingItemDecoration = ViewItemDecoration(this, R.layout.carbon_row_padding)
        paddingItemDecoration.setDrawBefore { it == 0 }
        paddingItemDecoration.setDrawAfter { it == adapter.itemCount - 1 }
        recycler.addItemDecoration(paddingItemDecoration)
    }
}